<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Using the Track object</title>
    
    <link rel="stylesheet" href="../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.3.0-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/logo.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Manipulating tracks" href="manipulate.html" />
    <link rel="prev" title="Loading a track" href="handeling.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="manipulate.html" title="Manipulating tracks"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="handeling.html" title="Loading a track"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">track v1.3.0-dev</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="using-the-track-object">
<h1>Using the Track object<a class="headerlink" href="#using-the-track-object" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="track.Track">
<em class="property">class </em><tt class="descclassname">track.</tt><tt class="descname">Track</tt><big>(</big><em>path</em>, <em>readonly=False</em>, <em>autosave=True</em>, <em>orig_path=None</em>, <em>orig_format=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track" title="Permalink to this definition">¶</a></dt>
<dd><p>The track object itself is iterable and will yield the name of all chromosomes.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/all_genes.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">t</span><span class="p">:</span> <span class="k">print</span> <span class="n">chrom</span>
    <span class="k">if</span> <span class="s">&#39;chrY&#39;</span> <span class="ow">in</span> <span class="n">t</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Male&#39;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">23</span><span class="p">:</span> <span class="k">print</span> <span class="s">&#39;Aneuploidy&#39;</span>
</pre></div>
</div>
</dd></dl>

<div class="section" id="track-attributes">
<h2>Track attributes<a class="headerlink" href="#track-attributes" title="Permalink to this headline">¶</a></h2>
<dl class="attribute">
<dt id="track.Track.fields">
<tt class="descclassname">Track.</tt><tt class="descname">fields</tt><a class="reference internal" href="../_modules/track.html#Track.fields"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.fields" title="Permalink to this definition">¶</a></dt>
<dd><p>A list the value types that each feature in the track will contain. For instance:</p>
<blockquote>
<div><tt class="docutils literal"><span class="pre">['start',</span> <span class="pre">'end',</span> <span class="pre">'name',</span> <span class="pre">'score',</span> <span class="pre">'strand']</span></tt></div></blockquote>
<p>Setting this attribute will influence the behaviour of all future read() and write() calls.</p>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.chromosomes">
<tt class="descclassname">Track.</tt><tt class="descname">chromosomes</tt><a class="reference internal" href="../_modules/track.html#Track.chromosomes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of all available chromosome. For instance:</p>
<blockquote>
<div><tt class="docutils literal"><span class="pre">['chr1,</span> <span class="pre">'chr2',</span> <span class="pre">'chr3',</span> <span class="pre">'chr4',</span> <span class="pre">'chr5',</span> <span class="pre">'chrC',</span> <span class="pre">'chrM']</span></tt></div></blockquote>
<p>You cannot set this attribute. To add new chromosomes, just <tt class="docutils literal"><span class="pre">write()</span></tt> to them.</p>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.info">
<tt class="descclassname">Track.</tt><tt class="descname">info</tt><a class="reference internal" href="../_modules/track.html#Track.info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.info" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary of meta data associated to the track (information like the source, etc). For instance:</p>
<p><tt class="docutils literal"><span class="pre">{'datatype':</span> <span class="pre">'signal',</span> <span class="pre">'source':</span> <span class="pre">'SGD',</span> <span class="pre">'orig_name':</span> <span class="pre">'splice_sites.bed'}</span></tt></p>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.name">
<tt class="descclassname">Track.</tt><tt class="descname">name</tt><a class="reference internal" href="../_modules/track.html#Track.name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Giving a name to your track is optional. The default name is the filename. This attribute is stored inside the <em>info</em> dictionary.</p>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.datatype">
<tt class="descclassname">Track.</tt><tt class="descname">datatype</tt><a class="reference internal" href="../_modules/track.html#Track.datatype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.datatype" title="Permalink to this definition">¶</a></dt>
<dd><p>Giving a datatype to your track is optional. The default datatype is <tt class="docutils literal"><span class="pre">None</span></tt>. Other possible datatypes are <tt class="docutils literal"><span class="pre">features</span></tt>, <tt class="docutils literal"><span class="pre">signal</span></tt> or <tt class="docutils literal"><span class="pre">relational</span></tt>. Changing the datatype imposes some conditions on the entries that the track contains. This attribute is stored inside the <em>info</em> dictionary.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">new</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">datatype</span> <span class="o">=</span> <span class="s">&#39;signal&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.assembly">
<tt class="descclassname">Track.</tt><tt class="descname">assembly</tt><a class="reference internal" href="../_modules/track.html#Track.assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Giving an assembly to your track is optional. However, if you set this variable for your track, you should input a valid assembly name such as &#8216;sacCer2&#8217;. Doing so will set the <em>chrmeta</em> attribute and rename all the chromosome to their canonical names if a correspondence is found. This attribute is also stored inside the <em>info</em> dictionary.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="n">track</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="s">&#39;tracks/genes.bed&#39;</span><span class="p">,</span> <span class="s">&#39;tracks/genes.sql&#39;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/genes.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="s">&#39;hg19&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.chrmeta">
<tt class="descclassname">Track.</tt><tt class="descname">chrmeta</tt><a class="reference internal" href="../_modules/track.html#Track.chrmeta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.chrmeta" title="Permalink to this definition">¶</a></dt>
<dd><p>Contains extra chromosomal meta data such as chromosome length information. <em>chrmeta</em> is a dictionary where each key is a chromosome names. For instance:</p>
<blockquote>
<div><tt class="docutils literal"><span class="pre">{'chr1':</span> <span class="pre">{'length':</span> <span class="pre">197195432},</span> <span class="pre">'chr2':</span> <span class="pre">{'length':</span> <span class="pre">129993255}}</span></tt></div></blockquote>
<p>You would hence use it like this:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">chrmeta</span><span class="p">[</span><span class="s">&#39;chr1&#39;</span><span class="p">][</span><span class="s">&#39;length&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Of course, genomic formats such as <tt class="docutils literal"><span class="pre">bed</span></tt> cannot store this kind of meta data. Hence, when loading tracks in these text formats, this information is lost once the track is closed.</p>
</dd></dl>

<dl class="attribute">
<dt id="track.Track.modified">
<tt class="descclassname">Track.</tt><tt class="descname">modified</tt><a class="reference internal" href="../_modules/track.html#Track.modified"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.modified" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean value which indicates if the track has been changed since it was opened. This value is set to False when you load a track and is set to True as soon, as you <tt class="docutils literal"><span class="pre">write</span></tt>, <tt class="docutils literal"><span class="pre">rename</span></tt> or <tt class="docutils literal"><span class="pre">remove</span></tt>. Changing the <tt class="docutils literal"><span class="pre">info</span></tt> or <tt class="docutils literal"><span class="pre">chrmeta</span></tt> attributes will also set this value to True.</p>
</dd></dl>

</div>
<div class="section" id="track-methods">
<h2>Track methods<a class="headerlink" href="#track-methods" title="Permalink to this headline">¶</a></h2>
<dl class="method">
<dt id="track.Track.read">
<tt class="descclassname">Track.</tt><tt class="descname">read</tt><big>(</big><em>selection=None</em>, <em>fields=None</em>, <em>order=''</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Read data from the track.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>selection</strong> &#8211; A chromosome name, or a dictionary specifying a region, see below.</li>
<li><strong>fields</strong> (<em>list of strings</em>) &#8211; is an optional list of fields which will influence the length of the tuples returned and the way in which the information is returned. The default is to read every field available for the given chromosome. If the <em>track.fields</em> attribute is set, that will be used.</li>
<li><strong>order</strong> (<em>comma-separated string</em>) &#8211; is an optional sublist of <em>fields</em> which will influence the order in which the tuples are yielded. By default results are not sorted.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a generator object yielding rows. A row can be referenced like a tuple or like a dictionary.</p>
</td>
</tr>
</tbody>
</table>
<p><em>selection</em> can be the name of a chromosome, in which case all the data on that chromosome will be returned.</p>
<p><em>selection</em> can be left empty, then the data from all chromosome is returned.</p>
<p><em>selection</em> can also be a dictionary specifying: regions, score intervals or strands. If you specify a region in which case only features contained in that region will be returned. But you can also input a tuple specifying a score interval in which case only features contained in those score boundaries will be returned. You can even specify a strand. The dictionary can contain one or several of these arguments. See code example for more details.</p>
<p>Adding the parameter <tt class="docutils literal"><span class="pre">'inclusion':'strict'</span></tt> to a region dictionary will return only features exactly contained inside the interval instead of features simply included in the interval. To combine multiple selections you can specify a list including chromosome names and region dictionaries. As expected, if such is the case, the joined data from those selections will be returned with an added <tt class="docutils literal"><span class="pre">chr</span></tt> field in front since the results may span several chromosomes.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;chr2&#39;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;chr3&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">,</span> <span class="s">&#39;strand&#39;</span><span class="p">])</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span><span class="s">&#39;chr2&#39;</span><span class="p">,</span><span class="s">&#39;chr3&#39;</span><span class="p">])</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">100</span><span class="p">})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">10000</span><span class="p">,</span> <span class="s">&#39;end&#39;</span><span class="p">:</span><span class="mi">15000</span><span class="p">})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">10000</span><span class="p">,</span> <span class="s">&#39;end&#39;</span><span class="p">:</span><span class="mi">15000</span><span class="p">,</span> <span class="s">&#39;inclusion&#39;</span><span class="p">:</span><span class="s">&#39;strict&#39;</span><span class="p">})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;strand&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;score&#39;</span><span class="p">:(</span><span class="mi">10</span><span class="p">,</span><span class="mi">100</span><span class="p">)})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">10000</span><span class="p">,</span> <span class="s">&#39;end&#39;</span><span class="p">:</span><span class="mi">15000</span><span class="p">,</span> <span class="s">&#39;strand&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="s">&#39;score&#39;</span><span class="p">:(</span><span class="mi">10</span><span class="p">,</span><span class="mi">100</span><span class="p">)})</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">read</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr5&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span> <span class="s">&#39;end&#39;</span><span class="p">:</span><span class="mi">200</span><span class="p">},</span> <span class="p">[</span><span class="s">&#39;strand&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">,</span> <span class="s">&#39;score&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.write">
<tt class="descclassname">Track.</tt><tt class="descname">write</tt><big>(</big><em>chromosome</em>, <em>data</em>, <em>fields=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write data to a genomic file. Will write many feature at once into a given chromosome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chromosome</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name of the chromosome on which one wants to write. For instance, if one is using the BED format this will become the first column, while if one is using the SQL format this will become the name of the table to be created.</li>
<li><strong>data</strong> (<em>an iteratable</em>) &#8211; must be an iterable object that yields tuples or rows of the correct length. As an example, the <tt class="docutils literal"><span class="pre">read</span></tt> function of this class produces such objects. <em>data</em> can have a <em>fields</em> attribute describing what the different elements of the tuple represent. <em>data</em> can also simply be a list of tuples.</li>
<li><strong>fields</strong> (<em>list of strings</em>) &#8211; is a parameter describing what the different elements in <em>data</em> represent. It is optional and is used only if <em>data</em> doesn&#8217;t already have a <tt class="docutils literal"><span class="pre">fields</span></tt> attribute.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None</p>
</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)])</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">def</span> <span class="nf">example_generator</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
            <span class="k">yield</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;chr2&#39;</span><span class="p">,</span> <span class="n">example_generator</span><span class="p">(),</span> <span class="n">fields</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;start&#39;</span><span class="p">,</span><span class="s">&#39;end&#39;</span><span class="p">,</span><span class="s">&#39;name&#39;</span><span class="p">])</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/new.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t2</span><span class="p">:</span>
    <span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/orig.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t1</span><span class="p">:</span>
        <span class="n">t1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="n">t2</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">))</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.save">
<tt class="descclassname">Track.</tt><tt class="descname">save</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Store the changes that were applied to the track on the disk. If the track was loaded from a text file such as &#8216;bed&#8217;, the file is rewritten with the changes included. If the track was loaded as an SQL file, the changes are committed to the database. Calling <tt class="docutils literal"><span class="pre">rollback</span></tt> will revert all changes to the track since the last call to <tt class="docutils literal"><span class="pre">save()</span></tt>. By default, when the track is closed, all changes are saved.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr19_gl000209_random&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.rollback">
<tt class="descclassname">Track.</tt><tt class="descname">rollback</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.rollback"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.rollback" title="Permalink to this definition">¶</a></dt>
<dd><p>Revert all changes to the track since the last call to <tt class="docutils literal"><span class="pre">save()</span></tt>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr19_gl000209_random&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="s">&#39;tmp/clean.bed&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">rollback</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.vacuum">
<tt class="descclassname">Track.</tt><tt class="descname">vacuum</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.vacuum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.vacuum" title="Permalink to this definition">¶</a></dt>
<dd><p>Rebuilds the database making it shrink in file size. This method is useful when, after having executed many inserts, updates, and deletes, the SQLite file is fragmented and full of empty space.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr19_gl000209_random&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">vaccum</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.close">
<tt class="descclassname">Track.</tt><tt class="descname">close</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.close"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close the current track. This method is useful when for some special reason you are not using the <tt class="docutils literal"><span class="pre">with</span> <span class="pre">...</span> <span class="pre">as`</span></tt> form for loading tracks.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr19_gl000209_random&#39;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.cursor">
<tt class="descclassname">Track.</tt><tt class="descname">cursor</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.cursor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.cursor" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new sqlite3 cursor object connected to the track database. You can use this attribute to make your own SQL queries and fetch the results. More information is available on the <a class="reference external" href="http://docs.python.org/library/sqlite3.html">sqlite3 documentation pages</a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A new sqlite3 cursor object</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">rpgenes</span><span class="p">:</span>
    <span class="n">cursor</span> <span class="o">=</span> <span class="n">rpgenes</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
    <span class="n">cursor</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s">&quot;select name from sqlite_master where type=&#39;table&#39;&quot;</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">cursor</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.export">
<tt class="descclassname">Track.</tt><tt class="descname">export</tt><big>(</big><em>path</em>, <em>format=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.export"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.export" title="Permalink to this definition">¶</a></dt>
<dd><p>Export the current track to a given format. A new file is created at the specified path. The current track object is unchanged</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the path to track file to create.</li>
<li><strong>format</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is an optional parameter specifying the format of the track to create when it cannot be guessed from the file extension.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None</p>
</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr19_gl000209_random&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="s">&#39;tmp/clean.bed&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">rollback</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.remove">
<tt class="descclassname">Track.</tt><tt class="descname">remove</tt><big>(</big><em>chromosome</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.remove"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.remove" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove data from a given chromosome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>chromosome</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name of the chromosome that one wishes to delete or a list of chromosomes to delete.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">remove</span><span class="p">([</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;chr2&#39;</span><span class="p">,</span> <span class="s">&#39;chr3&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.rename">
<tt class="descclassname">Track.</tt><tt class="descname">rename</tt><big>(</big><em>previous_name</em>, <em>new_name</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.rename"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.rename" title="Permalink to this definition">¶</a></dt>
<dd><p>Rename a chromosome from <em>previous_name</em> to <em>new_name</em></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>previous_name</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name of the chromosome that one wishes to rename.</li>
<li><strong>new_name</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name that that chromosome will now be referred by.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None.</p>
</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s">&#39;chr4&#39;</span><span class="p">,</span> <span class="s">&#39;chrIV&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.search">
<tt class="descclassname">Track.</tt><tt class="descname">search</tt><big>(</big><em>query_dict</em>, <em>fields=None</em>, <em>chromosome=None</em>, <em>exact_match=False</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Search for parameters inside your track. You can specify several parameters.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>selection</strong> &#8211; list of the fields you want to have in the result (to insure that all result will have the same number of columns)</li>
<li><strong>query_dict</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; A dictionary specifying keys and values to search for. See examples.</li>
<li><strong>chromosome</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; Optionally, the name of the chromosome on which one wants to search. If <tt class="docutils literal"><span class="pre">None</span></tt>, the search is performed on all chromosomes and every feature contains a new field specifying its chromosome.</li>
<li><strong>exact_match</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) &#8211; By default, will find all entries which contain the query. If set to <tt class="docutils literal"><span class="pre">True</span></tt>, will only find entries that exactly match the query.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a generator object yielding rows. A row can be referenced like a tuple or like a dictionary.</p>
</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search</span><span class="p">({</span><span class="s">&#39;gene_id&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">})</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search</span><span class="p">({</span><span class="s">&#39;gene_id&#39;</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="s">&#39;gene_name&#39;</span><span class="p">:</span><span class="s">&#39;YCCA3&#39;</span><span class="p">},</span> <span class="s">&#39;chr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.count">
<tt class="descclassname">Track.</tt><tt class="descname">count</tt><big>(</big><em>selection=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.count"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the number of features or entries in a given selection.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>selection</strong> &#8211; is the name of a chromosome, a list of chromosomes, a particular span or a list of spans. In other words, a value similar to the <em>selection</em> parameter of the <em>read</em> method. If left empty, will count every feature in a track</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">an integer.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">num</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">)</span>
    <span class="n">num</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">count</span><span class="p">([</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span><span class="s">&#39;chr2&#39;</span><span class="p">,</span><span class="s">&#39;chr3&#39;</span><span class="p">])</span>
    <span class="n">num</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">count</span><span class="p">({</span><span class="s">&#39;chr&#39;</span><span class="p">:</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="s">&#39;start&#39;</span><span class="p">:</span><span class="mi">10000</span><span class="p">,</span> <span class="s">&#39;end&#39;</span><span class="p">:</span><span class="mi">15000</span><span class="p">})</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.delete_fields">
<tt class="descclassname">Track.</tt><tt class="descname">delete_fields</tt><big>(</big><em>fields</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.delete_fields"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.delete_fields" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the given fields from all chromosomes. This is equivalent to dropping full columns in the database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fields</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; A list of fields such as <tt class="docutils literal"><span class="pre">['score','strand']</span></tt>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">fields</span>
    <span class="n">t</span><span class="o">.</span><span class="n">delete_fields</span><span class="p">([</span><span class="s">&#39;score&#39;</span><span class="p">,</span><span class="s">&#39;strand&#39;</span><span class="p">])</span>
    <span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">fields</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.load_chr_file">
<tt class="descclassname">Track.</tt><tt class="descname">load_chr_file</tt><big>(</big><em>path</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.load_chr_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.load_chr_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the <em>chrmeta</em> attribute of the track by loading a chromosome file. The chromosome file is structured as tab-separated text file containing two columns: the first specifies a chromosomes name and the second its length as an integer.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the file path to the chromosome file to load.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="track.Track.export_chr_file">
<tt class="descclassname">Track.</tt><tt class="descname">export_chr_file</tt><big>(</big><em>path</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.export_chr_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.export_chr_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Output the information contained in the <em>chrmeta</em> attribute into a plain text file. The chromosome file is structured as tab-separated text file containing two columns: the first specifies a chromosomes name and the second its length as an integer</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the file path to the chromosome file to create.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="track.Track.get_full_score_vector">
<tt class="descclassname">Track.</tt><tt class="descname">get_full_score_vector</tt><big>(</big><em>chromosome</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.get_full_score_vector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.get_full_score_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an iterable with as many elements as there are base pairs in the chromosomes specified by the <em>chromosome</em> parameter. Every element of the iterable is a float indicating the score at that position. If the track has no score associated, ones are inserted where features are present.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>chromosome</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name of the chromosome on which one wants to create a score vector from.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">an iterable yielding floats.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">new</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_full_score_vector</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.get_partial_score_vector">
<tt class="descclassname">Track.</tt><tt class="descname">get_partial_score_vector</tt><big>(</big><em>chromosome</em>, <em>start</em>, <em>end</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.get_partial_score_vector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.get_partial_score_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an iterable with as many elements as there are base pairs in the interval between <em>start</em> and <em>end</em>. Every element of the iterable is a float indicating the score at that position. If the track has no score associated, ones are inserted where features are present.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chromosome</strong> (<a class="reference external" href="http://docs.python.org/library/string.html#module-string" title="(in Python v2.7)"><em>string</em></a>) &#8211; is the name of the chromosome on which one wants to create a score vector from.</li>
<li><strong>start</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; The base pair position where scores will start being read from. Defaults to 0.</li>
<li><strong>end</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; The base pair position where scores will stop being read from. Defaults to the length of the chromosome.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">an iterable yielding floats.</p>
</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">new</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_partial_score_vector</span><span class="p">(</span><span class="s">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.ucsc_to_ensembl">
<tt class="descclassname">Track.</tt><tt class="descname">ucsc_to_ensembl</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.ucsc_to_ensembl"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.ucsc_to_ensembl" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert all entries of a track from the UCSC standard to the Ensembl standard effectively adding one to every start position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/example.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">ucsc_to_ensembl</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.ensembl_to_ucsc">
<tt class="descclassname">Track.</tt><tt class="descname">ensembl_to_ucsc</tt><big>(</big><big>)</big><a class="reference internal" href="../_modules/track.html#Track.ensembl_to_ucsc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.ensembl_to_ucsc" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts all entries of a track from the Ensembl standard to the UCSC standard effectively subtracting one from every start position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;tracks/rp_genes.bed&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">t</span><span class="o">.</span><span class="n">ensembl_to_ucsc</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.roman_to_integer">
<tt class="descclassname">Track.</tt><tt class="descname">roman_to_integer</tt><big>(</big><em>names=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.roman_to_integer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.roman_to_integer" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the name of all chromosomes from the roman numeral standard to the arabic numeral standard. For instance, &#8216;chrI&#8217; will become &#8216;chr1&#8217; while &#8216;chrII&#8217; will become &#8216;chr2&#8217;, etc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>names</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; an optional dictionary specifying how to translate particular cases. Example: <tt class="docutils literal"><span class="pre">{'chrM':'chrQ',</span> <span class="pre">'2micron':'chrR'}</span></tt></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">new</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">roman_to_integer</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="track.Track.integer_to_roman">
<tt class="descclassname">Track.</tt><tt class="descname">integer_to_roman</tt><big>(</big><em>names=None</em><big>)</big><a class="reference internal" href="../_modules/track.html#Track.integer_to_roman"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#track.Track.integer_to_roman" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the name of all chromosomes from the arabic numeral standard to the roman numeral standard. For instance, &#8216;chr1&#8217; will become &#8216;chrI&#8217; while &#8216;chr2&#8217; will become &#8216;chrII&#8217;, etc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>names</strong> (<a class="reference external" href="http://docs.python.org/library/stdtypes.html#dict" title="(in Python v2.7)"><em>dict</em></a>) &#8211; an optional dictionary specifying how to translate particular cases. Example: <tt class="docutils literal"><span class="pre">{'chrQ':'chrM',</span> <span class="pre">'chrR':'2micron'}</span></tt></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">None.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">track</span>
<span class="k">with</span> <span class="n">track</span><span class="o">.</span><span class="n">new</span><span class="p">(</span><span class="s">&#39;tmp/track.sql&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">roman_to_integer</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Using the Track object</a><ul>
<li><a class="reference internal" href="#track-attributes">Track attributes</a></li>
<li><a class="reference internal" href="#track-methods">Track methods</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="handeling.html"
                        title="previous chapter">Loading a track</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="manipulate.html"
                        title="next chapter">Manipulating tracks</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/track.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="manipulate.html" title="Manipulating tracks"
             >next</a> |</li>
        <li class="right" >
          <a href="handeling.html" title="Loading a track"
             >previous</a> |</li>
        <li><a href="../index.html">track v1.3.0-dev</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      Last updated on Jan 21, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>
  </body>
</html>